ckanext/archiver/controller.py

```diff
@@ -0,0 +1,93 @@
+import os
+import mimetypes
+import paste.fileapp
+from ckantoolkit import config
+
+import ckantoolkit as toolkit
+import ckan.logic as logic
+import ckan.lib.base as base
+import ckan.model as model
+import ckan.lib.uploader as uploader
+from ckan.common import _, request, c, response
+
+from ckanext.archiver.tasks import archiverDownloader
+import webob
+
+import logging
+log = logging.getLogger(__name__)
+
+NotFound = logic.NotFound
+NotAuthorized = logic.NotAuthorized
+get_action = logic.get_action
+abort = base.abort
+redirect = toolkit.redirect_to
+
+
+class ArchiverController(base.BaseController):
+
+    def archive_download(self, id, resource_id, filename=None):
+        '''
+        Provide a download by either redirecting the user to the url stored or
+        downloading the uploaded file from S3.
+        '''
+        context = {'model': model, 'session': model.Session,
+                   'user': c.user or c.author, 'auth_user_obj': c.userobj}
+
+        try:
+            rsc = get_action('resource_show')(context, {'id': resource_id})
+            get_action('package_show')(context, {'id': id})
+        except NotFound:
+            abort(404, _('Resource not found'))
+        except NotAuthorized:
+            abort(401, _('Unauthorized to read resource %s') % id)
+
+        if rsc.get('url_type') == 'upload':
+            upload = uploader.get_resource_uploader(rsc)
+            bucket_name = config.get('ckanext.s3filestore.aws_bucket_name')
```

Comment:
We ideally shouldn't make the assumption that we're always using S3, or else our changes won't be useful upstream. As discussed, let's piggyback on the regular resource download if possible.

Reply:
still spit balling, this should wrap the resource_download controller if possible